src/data/nat/basic.lean

```diff
@@ -1373,6 +1386,28 @@ by rw [nat.mul_add_mod, nat.mod_eq_of_lt h]
 lemma pred_eq_self_iff {n : ℕ} : n.pred = n ↔ n = 0 :=
 by { cases n; simp [(nat.succ_ne_self _).symm] }
 
+/-- Recursion principle based on the div and modulus decomposition against some base `b : ℕ`:
+given a map `C (b * k + r) → C (b * (k + 1) + r)` for each `r < b` and `k`,
+and the construction of `C r` for all `r < b`, one can construct `C n` for any `n : ℕ`.  -/
+def div_mod_induction {C : ℕ → Sort*} (n b : ℕ) (hb : b ≠ 0) (hr : ∀ r < b, C r)
```

Comment:
Does grepping for `∀ r < b, C r` in some form find any existing recursors? I'm pretty sure I've seen this

Reply:
```
$ rg "∀ \(?. < .\)?, C ."
src/scratch.lean
4:def nat.div_mod_induction {C : ℕ → Sort*} (n b : ℕ) (hb : b ≠ 0) (hr : ∀ r < b, C r)
```

Reply:
Maybe for `P` instead of `C`?

Reply:
```
src/ring_theory/artinian.lean
210:  (hgt : ∀ I, (∀ J < I, P J) → P I) (I : submodule R M) : P I :=

src/data/nat/basic.lean
388:theorem forall_lt_succ {P : ℕ → Prop} {n : ℕ} : (∀ m < n.succ, P m) ↔ (∀ m < n, P m) ∧ P n :=
```

Reply:
I had a look through the zulip archive, the closest I could find (via `(.) \(?.\)? ?→ ?\1 ?\(.*? ?\+ ?.*?\)` after erasing all the syntax highlighting `<span>`s) is:

* https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/.28.E2.88.80.20.28n.20.3A.20.E2.84.A4.29.2C.20P.20n.29.20.E2.88.A8.20.E2.88.80.20.28n.20.3A.20.E2.84.A4.29.2C.20.C2.ACP.20n/near/209235912

I still can't shake the feeling I've seen this before, but I'm a loss as to how to find out where.

Reply:
It might be named as some `shift` induction or `add` inductor. There might be a related one for int? I also have some deja vu, but can't find it. At least this one is now documented at the top of that `nat/basic.lean` file.